lib/rspec/support/fuzzy_matcher.rb

```diff
@@ -3,46 +3,100 @@ module Support
     # Provides a means to fuzzy-match between two arbitrary objects.
     # Understands array/hash nesting. Uses `===` or `==` to
     # perform the matching.
-    module FuzzyMatcher
-      # @api private
-      def self.values_match?(expected, actual)
+    class FuzzyMatcher
+      def initialize(expected, actual)
+        @path_to_not_matching = []
+        failure_message = match(expected, actual)
+        @matched = !failure_message
+        if failure_message
+          path = @path_to_not_matching.map { |key| "[#{key.inspect}]" }.join
+          inspected_path = "#{actual.inspect}#{path}"
+          @failure_message =
+            if failure_message.empty?
+              "#{inspected_path} failed to match"
+            else
+              "#{inspected_path} failed to match: #{failure_message}"
+            end
+        end
+      end
+
+      def matched?
+        @matched
+      end
+      attr_reader :failure_message
+
+      private
+
+      def match(expected, actual)
         if Hash === actual
-          return hashes_match?(expected, actual) if Hash === expected
+          return hashes_failure_message(expected, actual) if Hash === expected
         elsif Array === expected && Enumerable === actual && !(Struct === actual)
-          return arrays_match?(expected, actual.to_a)
+          return arrays_failure_message(expected, actual.to_a)
         end
 
-        return true if expected == actual
+        return if expected == actual
 
         begin
-          expected === actual
+          if expected === actual
+            return
+          else
+            if Matchers::BuiltIn::BaseMatcher === expected
```

Comment:
RSpec::Support must be usable w/o rspec-expectations loaded (which is the gem that defines `Matchers::BuiltIn::BaseMatcher`) so doing this kind of check here is problematic.